File: src/com/lushprojects/circuitjs1/client/ScopePlot2d.java
package com.lushprojects.circuitjs1.client;

import com.google.gwt.canvas.client.Canvas;
import com.google.gwt.canvas.dom.client.Context2d;

class ScopePlot2d {
    boolean enabled;
    boolean plotXY;
    double scaleX, scaleY;

    // X/Y axis plot indices (into scope.plots)
    int plotX = 0, plotY = 1;
    // Modulator plot indices (-1 = none)
    int plotBrightness = -1;
    int plotColorR = -1, plotColorG = -1, plotColorB = -1;
    // Auto-scales for modulator plots
    double scaleBrightness = 5, scaleR = 5, scaleG = 5, scaleB = 5;

    private Canvas imageCanvas;
    private Context2d imageContext;
    private int draw_ox, draw_oy;
    private int alphaCounter;

    Scope scope;

    ScopePlot2d(Scope scope) {
	this.scope = scope;
	scaleX = 5;
	scaleY = .1;
    }

    void allocImage() {
	if (!enabled)
	    return;
	if (imageCanvas == null) {
	    imageCanvas = Canvas.createIfSupported();
	    imageContext = imageCanvas.getContext2d();
	}
	imageCanvas.setWidth(scope.rect.width + "PX");
	imageCanvas.setHeight(scope.rect.height + "PX");
	imageCanvas.setCoordinateSpaceWidth(scope.rect.width);
	imageCanvas.setCoordinateSpaceHeight(scope.rect.height);
	clearView();
    }

    void clearView() {
	if (imageContext != null) {
	    if (scope.app.isPrintable()) {
		imageContext.setFillStyle("#eee");
	    } else {
		imageContext.setFillStyle("#111");
	    }
	    imageContext.fillRect(0, 0, scope.rect.width - 1, scope.rect.height - 1);
	}
	draw_ox = draw_oy = -1;
    }

    double calcGridPx(int width, int height) {
	int m = width < height ? width : height;
	return ((double)(m) / 2) / ((double)(scope.manDivisions) / 2 + 0.05);
    }

    // Draw a segment from (ox,oy) to (x2,y2) with the given color and alpha.
    void drawSegment(int ox, int oy, int x2, int y2, String color, double alpha) {
	if (alpha != 1.0)
	    imageContext.setGlobalAlpha(alpha);
	imageContext.setStrokeStyle(color);
	imageContext.beginPath();
	imageContext.moveTo(ox, oy);
	imageContext.lineTo(x2, y2);
	imageContext.stroke();
	if (alpha != 1.0)
	    imageContext.setGlobalAlpha(1.0);
    }

    void drawTo(int x2, int y2) {
	if (draw_ox == -1) {
	    draw_ox = x2;
	    draw_oy = y2;
	    return;
	}
	String color = scope.app.isPrintable() ? "#000000" : "#ffffff";
	drawSegment(draw_ox, draw_oy, x2, y2, color, 1.0);
	draw_ox = x2;
	draw_oy = y2;
    }

    void drawTo(int x2, int y2, String color, double alpha) {
	if (draw_ox == -1) {
	    draw_ox = x2;
	    draw_oy = y2;
	    return;
	}
	drawSegment(draw_ox, draw_oy, x2, y2, color, alpha);
	draw_ox = x2;
	draw_oy = y2;
    }

    // Map a value in [-scale, +scale] to [0, 255] (bipolar).
    int mapBipolar(double value, double scale) {
	return (int) Math.max(0, Math.min(255, (value / scale + 1) * 127.5));
    }

    // Compute the draw color from R/G/B modulator plots.
    String computeColor() {
	if (plotColorR < 0 && plotColorG < 0 && plotColorB < 0)
	    return scope.app.isPrintable() ? "#000000" : "#ffffff";
	int r = 0, g = 0, b = 0;
	if (plotColorR >= 0 && plotColorR < scope.plots.size()) {
	    double rv = scope.plots.get(plotColorR).lastValue;
	    while (Math.abs(rv) > scaleR) scaleR *= 2;
	    r = mapBipolar(rv, scaleR);
	}
	if (plotColorG >= 0 && plotColorG < scope.plots.size()) {
	    double gv = scope.plots.get(plotColorG).lastValue;
	    while (Math.abs(gv) > scaleG) scaleG *= 2;
	    g = mapBipolar(gv, scaleG);
	}
	if (plotColorB >= 0 && plotColorB < scope.plots.size()) {
	    double bv = scope.plots.get(plotColorB).lastValue;
	    while (Math.abs(bv) > scaleB) scaleB *= 2;
	    b = mapBipolar(bv, scaleB);
	}
	return "rgb(" + r + "," + g + "," + b + ")";
    }

    // Compute draw alpha from the brightness modulator plot (0 = off, 1 = full).
    double computeAlpha() {
	if (plotBrightness < 0 || plotBrightness >= scope.plots.size())
	    return 1.0;
	double bv = Math.abs(scope.plots.get(plotBrightness).lastValue);
	while (bv > scaleBrightness) scaleBrightness *= 2;
	return scaleBrightness > 0 ? bv / scaleBrightness : 0;
    }

    // Clamp plotX/plotY to valid range.
    int validPlotIndex(int idx, int defaultIdx) {
	if (scope.plots.size() == 0) return 0;
	if (idx < 0 || idx >= scope.plots.size()) return Math.min(defaultIdx, scope.plots.size() - 1);
	return idx;
    }

    void timeStep() {
	if (imageContext == null || scope.plots.size() < 1)
	    return;
	int px = validPlotIndex(plotX, 0);
	int py = validPlotIndex(plotY, Math.min(1, scope.plots.size() - 1));
	double v = scope.plots.get(px).lastValue;
	double yval = scope.plots.get(py).lastValue;
	int x, y;
	if (!scope.isManualScale()) {
	    boolean newscale = false;
	    while (v > scaleX || v < -scaleX) { scaleX *= 2; newscale = true; }
	    while (yval > scaleY || yval < -scaleY) { scaleY *= 2; newscale = true; }
	    if (newscale)
		clearView();
	    double xa = v / scaleX;
	    double ya = yval / scaleY;
	    x = (int)(scope.rect.width  * (1 + xa) * .499);
	    y = (int)(scope.rect.height * (1 - ya) * .499);
	} else {
	    double gridPx = calcGridPx(scope.rect.width, scope.rect.height);
	    x = (int)(scope.rect.width  * .499 + (v    / scope.plots.get(px).manScale) * gridPx + gridPx * scope.manDivisions * (double)(scope.plots.get(px).manVPosition) / (double)(Scope.V_POSITION_STEPS));
	    y = (int)(scope.rect.height * .499 - (yval / scope.plots.get(py).manScale) * gridPx - gridPx * scope.manDivisions * (double)(scope.plots.get(py).manVPosition) / (double)(Scope.V_POSITION_STEPS));
	}
	String color = computeColor();
	double alpha = computeAlpha();
	drawTo(x, y, color, alpha);
    }

    void maxScale() {
	double x = 1e-8;
	scope.scale[Scope.UNITS_V]    *= x;
	scope.scale[Scope.UNITS_A]    *= x;
	scope.scale[Scope.UNITS_OHMS] *= x;
	scope.scale[Scope.UNITS_W]    *= x;
	scope.scale[Scope.UNITS_C]    *= x;
	scaleX *= x;
	scaleY *= x;
	scaleBrightness *= x;
	scaleR *= x;
	scaleG *= x;
	scaleB *= x;
    }

    void draw(Graphics g) {
	if (imageContext == null)
	    return;
	g.context.save();
	g.context.translate(scope.rect.x, scope.rect.y);
	g.clipRect(0, 0, scope.rect.width, scope.rect.height);

	alphaCounter++;
	if (alphaCounter > 2) {
	    alphaCounter = 0;
	    imageContext.setGlobalAlpha(0.01);
	    if (scope.app.isPrintable()) {
		imageContext.setFillStyle("#ffffff");
	    } else {
		imageContext.setFillStyle("#000000");
	    }
	    imageContext.fillRect(0, 0, scope.rect.width, scope.rect.height);
	    imageContext.setGlobalAlpha(1.0);
	}

	g.context.drawImage(imageContext.getCanvas(), 0.0, 0.0);
	g.setColor(CircuitElm.whiteColor);
	g.fillOval(draw_ox - 2, draw_oy - 2, 5, 5);
	g.setColor(CircuitElm.positiveColor);
	g.drawLine(0, scope.rect.height / 2, scope.rect.width - 1, scope.rect.height / 2);
	if (!plotXY)
	    g.setColor(Color.yellow);
	g.drawLine(scope.rect.width / 2, 0, scope.rect.width / 2, scope.rect.height - 1);
	if (scope.isManualScale()) {
	    double gridPx = calcGridPx(scope.rect.width, scope.rect.height);
	    g.setColor("#404040");
	    for (int i = -scope.manDivisions; i <= scope.manDivisions; i++) {
		if (i != 0)
		    g.drawLine((int)(gridPx * i) + scope.rect.width / 2, 0, (int)(gridPx * i) + scope.rect.width / 2, scope.rect.height);
		g.drawLine(0, (int)(gridPx * i) + scope.rect.height / 2, scope.rect.width, (int)(gridPx * i) + scope.rect.height / 2);
	    }
	}
	scope.overlays.textY = 10;
	g.setColor(CircuitElm.whiteColor);
	if (scope.text != null)
	    scope.drawInfoText(g, scope.text);
	int px = validPlotIndex(plotX, 0);
	int py = validPlotIndex(plotY, Math.min(1, scope.plots.size() - 1));
	if (scope.showScale && scope.plots.size() >= 1 && scope.isManualScale() && px < scope.plots.size() && py < scope.plots.size()) {
	    ScopePlot spx = scope.plots.get(px);
	    ScopePlot spy = scope.plots.get(py);
	    scope.drawInfoText(g, "X=" + spx.getUnitText(spx.manScale) + "/div, Y=" + spy.getUnitText(spy.manScale) + "/div");
	}
	g.context.restore();
	scope.drawSettingsWheel(g);
	if (!scope.app.dialogIsShowing() && scope.rect.contains(scope.app.mouse.mouseCursorX, scope.app.mouse.mouseCursorY) && scope.plots.size() >= 1 && px < scope.plots.size() && py < scope.plots.size()) {
	    double gridPx = calcGridPx(scope.rect.width, scope.rect.height);
	    String[] info = new String[2];
	    ScopePlot spx = scope.plots.get(px);
	    ScopePlot spy = scope.plots.get(py);
	    double xValue, yValue;
	    if (scope.isManualScale()) {
		xValue = spx.manScale * ((double)(scope.app.mouse.mouseCursorX - scope.rect.x - scope.rect.width  / 2) / gridPx - scope.manDivisions * spx.manVPosition / (double)(Scope.V_POSITION_STEPS));
		yValue = spy.manScale * ((double)(-scope.app.mouse.mouseCursorY + scope.rect.y + scope.rect.height / 2) / gridPx - scope.manDivisions * spy.manVPosition / (double)(Scope.V_POSITION_STEPS));
	    } else {
		xValue =  ((double)(scope.app.mouse.mouseCursorX - scope.rect.x) / (0.499 * (double)(scope.rect.width))  - 1.0) * scaleX;
		yValue = -((double)(scope.app.mouse.mouseCursorY - scope.rect.y) / (0.499 * (double)(scope.rect.height)) - 1.0) * scaleY;
	    }
	    info[0] = spx.getUnitText(xValue);
	    info[1] = spy.getUnitText(yValue);
	    scope.drawCursorInfo(g, info, 2, scope.app.mouse.mouseCursorX, true);
	}
    }
}

File: ts/ScopePlot2d.ts
/*
    Copyright (C) Paul Falstad and Iain Sharp

    This file is part of CircuitJS1.

    CircuitJS1 is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 2 of the License, or
    (at your option) any later version.

    CircuitJS1 is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CircuitJS1.  If not, see <http://www.gnu.org/licenses/>.
*/

import { CircuitElm } from "./CircuitElm";
import { Color } from "./Color";
import { Graphics } from "./Graphics";
import { Scope } from "./Scope";

export class ScopePlot2d {
    enabled: boolean = false;
    plotXY: boolean = false;
    scaleX: number;
    scaleY: number;

    // X/Y axis plot indices (into scope.plots)
    plotX: number = 0;
    plotY: number = 1;
    // Modulator plot indices (-1 = none)
    plotBrightness: number = -1;
    plotColorR: number = -1;
    plotColorG: number = -1;
    plotColorB: number = -1;
    // Auto-scales for modulator plots
    scaleBrightness: number = 5;
    scaleR: number = 5;
    scaleG: number = 5;
    scaleB: number = 5;

    private imageCanvas: HTMLCanvasElement | null = null;
    private imageContext: CanvasRenderingContext2D | null = null;
    private draw_ox: number = -1;
    private draw_oy: number = -1;
    private alphaCounter: number = 0;
    trailPersistence: number = 0;
    static readonly DEFAULT_TRAIL_PERSISTENCE = 0;
    lastTrailSimTime: number = -1;

    scope: Scope;

    constructor(scope: Scope) {
        this.scope = scope;
        this.scaleX = 5;
        this.scaleY = .1;
    }

    allocImage(): void {
        if (!this.enabled)
            return;
        if (this.imageCanvas === null) {
            this.imageCanvas = document.createElement("canvas");
            this.imageContext = this.imageCanvas.getContext("2d");
        }
        this.imageCanvas.width  = this.scope.rect.width;
        this.imageCanvas.height = this.scope.rect.height;
        this.clearView();
    }

    clearView(): void {
        if (this.imageContext !== null) {
            this.imageContext.fillStyle = this.scope.app.isPrintable() ? "#eee" : "#111";
            this.imageContext.fillRect(0, 0, this.scope.rect.width - 1, this.scope.rect.height - 1);
        }
        this.draw_ox = this.draw_oy = -1;
    }

    calcGridPx(width: number, height: number): number {
        const m = Math.min(width, height);
        return (m / 2) / (this.scope.manDivisions / 2 + 0.05);
    }

    // Draw a segment from (ox,oy) to (x2,y2) with the given color and alpha.
    drawSegment(ox: number, oy: number, x2: number, y2: number, color: string, alpha: number): void {
        if (alpha !== 1.0)
            this.imageContext!.globalAlpha = alpha;
        this.imageContext!.strokeStyle = color;
        this.imageContext!.beginPath();
        this.imageContext!.moveTo(ox, oy);
        this.imageContext!.lineTo(x2, y2);
        this.imageContext!.stroke();
        if (alpha !== 1.0)
            this.imageContext!.globalAlpha = 1.0;
    }

    drawTo(x2: number, y2: number, color?: string, alpha?: number): void {
        if (this.draw_ox === -1) {
            this.draw_ox = x2;
            this.draw_oy = y2;
            return;
        }
        const c = color ?? (this.scope.app.isPrintable() ? "#000000" : "#ffffff");
        this.drawSegment(this.draw_ox, this.draw_oy, x2, y2, c, alpha ?? 1.0);
        this.draw_ox = x2;
        this.draw_oy = y2;
    }

    // Compute the draw color from R/G/B modulator plots.
    computeColor(): string {
        if (this.plotColorR < 0 && this.plotColorG < 0 && this.plotColorB < 0)
            return this.scope.app.isPrintable() ? "#000000" : "#ffffff";
        let r = 0, g = 0, b = 0;
        if (this.plotColorR >= 0 && this.plotColorR < this.scope.plots.length) {
            const rv = this.scope.plots[this.plotColorR].lastValue;
            while (rv > this.scaleR) this.scaleR *= 2;
            r = Math.trunc(Math.max(0, Math.min(255, (rv / this.scaleR) * 255)));
        }
        if (this.plotColorG >= 0 && this.plotColorG < this.scope.plots.length) {
            const gv = this.scope.plots[this.plotColorG].lastValue;
            while (gv > this.scaleG) this.scaleG *= 2;
            g = Math.trunc(Math.max(0, Math.min(255, (gv / this.scaleG) * 255)));
        }
        if (this.plotColorB >= 0 && this.plotColorB < this.scope.plots.length) {
            const bv = this.scope.plots[this.plotColorB].lastValue;
            while (bv > this.scaleB) this.scaleB *= 2;
            b = Math.trunc(Math.max(0, Math.min(255, (bv / this.scaleB) * 255)));
        }
        return "rgb(" + r + "," + g + "," + b + ")";
    }

    // Compute draw alpha from the brightness modulator plot (0 = off, 1 = full).
    computeAlpha(): number {
        if (this.plotBrightness < 0 || this.plotBrightness >= this.scope.plots.length)
            return 1.0;
        const bv = Math.abs(this.scope.plots[this.plotBrightness].lastValue);
        while (bv > this.scaleBrightness) this.scaleBrightness *= 2;
        return this.scaleBrightness > 0 ? bv / this.scaleBrightness : 0;
    }

    // Clamp plotX/plotY to valid range.
    validPlotIndex(idx: number, defaultIdx: number): number {
        if (this.scope.plots.length === 0) return 0;
        if (idx < 0 || idx >= this.scope.plots.length) return Math.min(defaultIdx, this.scope.plots.length - 1);
        return idx;
    }

    timeStep(): void {
        if (this.imageContext === null || this.scope.plots.length < 1)
            return;
        const px = this.validPlotIndex(this.plotX, 0);
        const py = this.validPlotIndex(this.plotY, Math.min(1, this.scope.plots.length - 1));
        const v    = this.scope.plots[px].lastValue;
        const yval = this.scope.plots[py].lastValue;
        let x: number, y: number;
        if (!this.scope.isManualScale()) {
            let newscale = false;
            while (v > this.scaleX || v < -this.scaleX)    { this.scaleX *= 2; newscale = true; }
            while (yval > this.scaleY || yval < -this.scaleY) { this.scaleY *= 2; newscale = true; }
            if (newscale)
                this.clearView();
            x = Math.trunc(this.scope.rect.width  * (1 + v    / this.scaleX) * .499);
            y = Math.trunc(this.scope.rect.height * (1 - yval / this.scaleY) * .499);
        } else {
            const gridPx = this.calcGridPx(this.scope.rect.width, this.scope.rect.height);
            x = Math.trunc(this.scope.rect.width  * .499 + (v    / this.scope.plots[px].manScale) * gridPx + gridPx * this.scope.manDivisions * this.scope.plots[px].manVPosition  / Scope.V_POSITION_STEPS);
            y = Math.trunc(this.scope.rect.height * .499 - (yval / this.scope.plots[py].manScale) * gridPx - gridPx * this.scope.manDivisions * this.scope.plots[py].manVPosition / Scope.V_POSITION_STEPS);
        }
        this.drawTo(x, y, this.computeColor(), this.computeAlpha());
    }

    maxScale(): void {
        const x = 1e-8;
        this.scope.scale[Scope.UNITS_V]    *= x;
        this.scope.scale[Scope.UNITS_A]    *= x;
        this.scope.scale[Scope.UNITS_OHMS] *= x;
        this.scope.scale[Scope.UNITS_W]    *= x;
        this.scope.scale[Scope.UNITS_C]    *= x;
        this.scaleX            *= x;
        this.scaleY            *= x;
        this.scaleBrightness   *= x;
        this.scaleR            *= x;
        this.scaleG            *= x;
        this.scaleB            *= x;
    }

    draw(g: Graphics): void {
        if (this.imageContext === null)
            return;
        g.context.save();
        g.context.translate(this.scope.rect.x, this.scope.rect.y);
        g.clipRect(0, 0, this.scope.rect.width, this.scope.rect.height);

        this.alphaCounter++;
        if (this.alphaCounter > 2) {
            this.alphaCounter = 0;
            let fadeAlpha: number;
            if (this.trailPersistence <= 0) {
                fadeAlpha = 0.01;
            } else {
                // Sim-time exponential fade; time constant = trailPersistence * maxTimeStep (seconds).
                // Don't advance lastTrailSimTime until fadeAlpha is large enough; otherwise sub-pixel
                // alphas have no effect on an 8-bit canvas and the trail never fades at low speed.
                if (this.lastTrailSimTime < 0 || this.scope.sim.t < this.lastTrailSimTime)
                    this.lastTrailSimTime = this.scope.sim.t;
                const elapsed   = this.scope.sim.t - this.lastTrailSimTime;
                const timeConst = this.trailPersistence * this.scope.sim.maxTimeStep;
                fadeAlpha = 1.0 - Math.exp(-elapsed / timeConst);
                if (fadeAlpha >= 3.0 / 255)
                    this.lastTrailSimTime = this.scope.sim.t;
                else
                    fadeAlpha = 0;
            }
            if (fadeAlpha > 0) {
                this.imageContext.globalAlpha = fadeAlpha;
                this.imageContext.fillStyle = this.scope.app.isPrintable() ? "#ffffff" : "#000000";
                this.imageContext.fillRect(0, 0, this.scope.rect.width, this.scope.rect.height);
                this.imageContext.globalAlpha = 1.0;
            }
        }

        g.context.drawImage(this.imageCanvas!, 0, 0);
        g.setColor(CircuitElm.whiteColor);
        g.fillOval(this.draw_ox - 2, this.draw_oy - 2, 5, 5);
        g.setColor(CircuitElm.positiveColor);
        g.drawLine(0, this.scope.rect.height / 2, this.scope.rect.width - 1, this.scope.rect.height / 2);
        if (!this.plotXY)
            g.setColor(Color.yellow);
        g.drawLine(this.scope.rect.width / 2, 0, this.scope.rect.width / 2, this.scope.rect.height - 1);
        if (this.scope.isManualScale()) {
            const gridPx = this.calcGridPx(this.scope.rect.width, this.scope.rect.height);
            g.setColor("#404040");
            for (let i = -this.scope.manDivisions; i <= this.scope.manDivisions; i++) {
                if (i !== 0)
                    g.drawLine(Math.trunc(gridPx * i) + this.scope.rect.width  / 2, 0,
                               Math.trunc(gridPx * i) + this.scope.rect.width  / 2, this.scope.rect.height);
                g.drawLine(0,                          Math.trunc(gridPx * i) + this.scope.rect.height / 2,
                           this.scope.rect.width,      Math.trunc(gridPx * i) + this.scope.rect.height / 2);
            }
        }
        this.scope.overlays.textY = 10;
        g.setColor(CircuitElm.whiteColor);
        if (this.scope.text != null)
            this.scope.drawInfoText(g, this.scope.text);
        const px = this.validPlotIndex(this.plotX, 0);
        const py = this.validPlotIndex(this.plotY, Math.min(1, this.scope.plots.length - 1));
        if (this.scope.showScale && this.scope.plots.length >= 1 && this.scope.isManualScale() &&
                px < this.scope.plots.length && py < this.scope.plots.length) {
            const spx = this.scope.plots[px];
            const spy = this.scope.plots[py];
            this.scope.drawInfoText(g, "X=" + spx.getUnitText(spx.manScale) + "/div, Y=" + spy.getUnitText(spy.manScale) + "/div");
        }
        g.context.restore();
        this.scope.drawSettingsWheel(g);
        if (!this.scope.app.modalDialogIsShowing() &&
                this.scope.rect.contains(this.scope.app.mouse.mouseCursorX, this.scope.app.mouse.mouseCursorY) &&
                this.scope.plots.length >= 1 && px < this.scope.plots.length && py < this.scope.plots.length) {
            const gridPx = this.calcGridPx(this.scope.rect.width, this.scope.rect.height);
            const info: string[] = new Array(2);
            const spx = this.scope.plots[px];
            const spy = this.scope.plots[py];
            let xValue: number, yValue: number;
            if (this.scope.isManualScale()) {
                xValue =  spx.manScale * ((this.scope.app.mouse.mouseCursorX - this.scope.rect.x - this.scope.rect.width  / 2) / gridPx - this.scope.manDivisions * spx.manVPosition / Scope.V_POSITION_STEPS);
                yValue =  spy.manScale * ((-this.scope.app.mouse.mouseCursorY + this.scope.rect.y + this.scope.rect.height / 2) / gridPx - this.scope.manDivisions * spy.manVPosition / Scope.V_POSITION_STEPS);
            } else {
                xValue =  ((this.scope.app.mouse.mouseCursorX - this.scope.rect.x) / (0.499 * this.scope.rect.width)  - 1.0) * this.scaleX;
                yValue = -((this.scope.app.mouse.mouseCursorY - this.scope.rect.y) / (0.499 * this.scope.rect.height) - 1.0) * this.scaleY;
            }
            info[0] = spx.getUnitText(xValue);
            info[1] = spy.getUnitText(yValue);
            this.scope.drawCursorInfo(g, info, 2, this.scope.app.mouse.mouseCursorX, true);
        }
    }
}
